src/utransport.rs

```diff
@@ -166,20 +326,258 @@ impl UTransport for UPClientZenoh {
+            "Invalid uAttributes",
+        ))?;
+
+        // Check the type of UAttributes (Publish / Request / Response)
+        match attributes
+            .type_
+            .enum_value()
+            .map_err(|_| UStatus::fail_with_code(UCode::INTERNAL, "Unable to parse type"))?
+        {
+            UMessageType::UMESSAGE_TYPE_PUBLISH => {
+                Validators::Publish
+                    .validator()
+                    .validate(&attributes)
+                    .map_err(|e| {
+                        UStatus::fail_with_code(
+                            UCode::INVALID_ARGUMENT,
+                            format!("Wrong Publish UAttributes {e:?}"),
+                        )
+                    })?;
+                // Get Zenoh key
+                let topic = attributes.clone().sink;
+                let zenoh_key = UPClientZenoh::to_zenoh_key_string(&topic)?;
+                // Send Publish
+                self.send_publish(&zenoh_key, payload, attributes).await
+            }
+            UMessageType::UMESSAGE_TYPE_REQUEST => {
+                Validators::Request
+                    .validator()
+                    .validate(&attributes)
+                    .map_err(|e| {
+                        UStatus::fail_with_code(
+                            UCode::INVALID_ARGUMENT,
+                            format!("Wrong Request UAttributes {e:?}"),
+                        )
+                    })?;
+                // Get Zenoh key
+                let topic = attributes.clone().sink;
+                let zenoh_key = UPClientZenoh::to_zenoh_key_string(&topic)?;
+                // Send Request
+                self.send_request(&zenoh_key, payload, attributes).await
+            }
+            UMessageType::UMESSAGE_TYPE_RESPONSE => {
+                Validators::Response
+                    .validator()
+                    .validate(&attributes)
+                    .map_err(|e| {
+                        UStatus::fail_with_code(
+                            UCode::INVALID_ARGUMENT,
+                            format!("Wrong Response UAttributes {e:?}"),
+                        )
+                    })?;
+                // Get Zenoh key
+                let topic = attributes.clone().source;
+                let zenoh_key = UPClientZenoh::to_zenoh_key_string(&topic)?;
+                // Send Response
+                self.send_response(&zenoh_key, payload, attributes).await
+            }
+            UMessageType::UMESSAGE_TYPE_UNSPECIFIED => Err(UStatus::fail_with_code(
+                UCode::INVALID_ARGUMENT,
+                "Wrong Message type in UAttributes",
+            )),
+        }
+    }
+
+    async fn receive(&self, _topic: UUri) -> Result<UMessage, UStatus> {
+        Err(UStatus::fail_with_code(
+            UCode::UNIMPLEMENTED,
+            "Not implemented",
+        ))
+    }
+
+    async fn register_listener(
+        &self,
+        topic: UUri,
```

Comment:
I'm also not seeing in here special logic to allow us to register for _all three_ of Request, Response, and Publish message types of the UUri matches the format [we discussed here](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2#discussion_r1474661323).

Is it understood that it is a requirement of the uStreamer on all `up-client-foo-rust` to be able to register to all three types of messages transparently via `UTransport::register_listener()` passing that UUri we agreed upon with only UAuthority filled out?

Reply:
Reading the a little further on how you implemented `UTransport::send()`, with it dispatching to the three types under the hood and how `send_request()` is implemented, I think the requirement could be met by having a special path through this function that checks if we have only UAuthority filled out, and if so, does the contents of all three branches of the if/else if/else below.

Would still have to have a way to have the UAuthority alone serialized, which as I [wrote in this comment](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2/commits/689bb50c22271f1f5db2aab9d005824aec865d19#r1501858724) should entail that work being pushed into `up-rust` so that it is done consistently across all `up-client-foo-rust`.

Does that make sense?

Reply:
Additionally, it should _not_ be possible to register for Response or Request type messages through `UTransport::register_listener()` at all unless we are in the uStreamer use-case.

So roughly speaking it should probably be:
* Is the UUri the special uStreamer case?
   * Yes => run all three below branches to register for all of Publish, Request, Response
   * No => validate it's a good one to register for Publish messages

Does that make sense?

Reply:
> I'm also not seeing in here special logic to allow us to register for all three of Request, Response, and Publish message types of the UUri matches the format https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2#discussion_r1474661323.

My bad. I'll add the special logic here.

> I think the requirement could be met by having a special path through this function that checks if we have only UAuthority filled out, and if so, does the contents of all three branches of the if/else if/else below.

Agree. But I think we only need to register Publish & Request, while listening to Response is only needed by the side which sends the request.

> Additionally, it should not be possible to register for Response or Request type messages through UTransport::register_listener() at all unless we are in the uStreamer use-case.

I don't think we should do that since `invoke_method` and `register_rpc_listener()` also call `send` and `register_listner()`. What in my mind is that everyone can use `send` and `register_listner()`, but they can simplify the process in the RPC scenario if they use `invoke_method` and `register_rpc_listener()`.
Let's talk about this later in today meeting.

Reply:
> Agree. But I think we only need to register Publish & Request, while listening to Response is only needed by the side which sends the request.

The uStreamer will be sending Requests as well though. If the uStreamer gets a Request over sommR, and needs to "forward" that request onto Zenoh, the uStreamer will need a way to listen for the Response to send it back over sommR.

> Let's talk about this later in today meeting.

Sure, sounds good